File: worker/src/domain/entities/anomaly.py
"""Anomaly domain entity."""

from dataclasses import dataclass
from datetime import datetime


@dataclass
class AnomalyEntity:
    """Represents an anomaly detected during ETL processing.

    Maps to the ``anomalies`` table in Prisma schema.
    """

    id: str
    dataset_id: str
    column: str
    row: int | None
    type: str  # "MISSING_VALUE" | "OUTLIER" | "FORMAT_ERROR" | "DUPLICATE"
    description: str
    original_value: str | None
    suggested_value: str | None
    status: str  # "PENDING" | "RESOLVED"
    created_at: datetime

    @classmethod
    def create(
        cls,
        *,
        id: str,
        dataset_id: str,
        column: str,
        row: int | None,
        anomaly_type: str,
        description: str,
        original_value: str | None = None,
        suggested_value: str | None = None,
    ) -> "AnomalyEntity":
        """Create a new anomaly entity."""
        return cls(
            id=id,
            dataset_id=dataset_id,
            column=column,
            row=row,
            type=anomaly_type,
            description=description,
            original_value=original_value,
            suggested_value=suggested_value,
            status="PENDING",
            created_at=datetime.utcnow(),
        )

File: api/src/domain/entities/Anomaly.ts
export type AnomalyType = 'MISSING_VALUE' | 'OUTLIER' | 'FORMAT_ERROR' | 'DUPLICATE';
export type AnomalyStatus = 'PENDING' | 'RESOLVED';
export type DecisionAction = 'APPROVED' | 'CORRECTED' | 'DISCARDED';

export interface AnomalyDecision {
  id: string;
  anomalyId: string;
  action: DecisionAction;
  correction: string | null;
  userId: string;
  createdAt: Date;
}

export interface AnomalyProps {
  id: string;
  datasetId: string;
  column: string;
  row: number | null;
  type: AnomalyType;
  description: string;
  originalValue: string | null;
  suggestedValue: string | null;
  status: AnomalyStatus;
  createdAt: Date;
  updatedAt: Date;
  decision: AnomalyDecision | null;
}

/**
 * Anomaly domain entity.
 * Represents a data quality issue detected in a dataset that requires
 * a human decision (Human-in-the-Loop).
 */
export class Anomaly {
  private constructor(private readonly props: AnomalyProps) {}

  static create(
    props: Omit<AnomalyProps, 'status' | 'createdAt' | 'updatedAt' | 'decision'>
  ): Anomaly {
    return new Anomaly({
      ...props,
      status: 'PENDING',
      createdAt: new Date(),
      updatedAt: new Date(),
      decision: null,
    });
  }

  static reconstitute(props: AnomalyProps): Anomaly {
    return new Anomaly(props);
  }

  get id(): string {
    return this.props.id;
  }

  get datasetId(): string {
    return this.props.datasetId;
  }

  get column(): string {
    return this.props.column;
  }

  get row(): number | null {
    return this.props.row;
  }

  get type(): AnomalyType {
    return this.props.type;
  }

  get description(): string {
    return this.props.description;
  }

  get originalValue(): string | null {
    return this.props.originalValue;
  }

  get suggestedValue(): string | null {
    return this.props.suggestedValue;
  }

  get status(): AnomalyStatus {
    return this.props.status;
  }

  get createdAt(): Date {
    return this.props.createdAt;
  }

  get updatedAt(): Date {
    return this.props.updatedAt;
  }

  get decision(): AnomalyDecision | null {
    return this.props.decision;
  }

  resolve(decision: AnomalyDecision): void {
    this.props.status = 'RESOLVED';
    this.props.decision = decision;
    this.props.updatedAt = new Date();
  }
}
